Fix relative imports, transformerConfig, configMerge; Allow functional config

cpiber opened the pull request:
- Fix relative imports: `rollup-plugin-external` had a problem and marked file imports as external
- Fix transformerConfig: At least with projects, jest passes the matching config directly via transformerConfig, not tested whether the current solution is necessary at all
- Fix configMerge: Merging the three configs should be done in three steps. Before, it was possible for the config from the file to override the config from jest (instead of merging). The plugin transform also wasn't applied to the config from the file
- Allow functional config: Rollup allows config files to export a function of the form `args => config`, added support for that. A special key `args` has been added to support arguments that aren't accepted in inputOptions (but on CLI). The function is passed `{ ...args, ...options }`.
- Allow disabling cache: With the new key `useCache`, caching of entry file can be prevented, useful f.e. in combination with @rollup/plugin-typescript (which doesn't operate on the cached version)

cpiber committed:
Accept transform config from given

cpiber committed:
Support function config

cpiber committed:
Fix relative imports being treated as external

cpiber committed:
Remove pnpm-lock.yml

cpiber committed:
Use separate key for function config, fix merge

concatMerge should have been applied separately to avoid overriding fields
Provide separate args key for options that aren't accepted in inputOptions

cpiber committed:
Test for config function

cpiber committed:
Allow disabling cache

cpiber commented:
> Fix relative imports: rollup-plugin-external had a problem and marked file imports as external

Actually, could you please give a statement on that? In hindsight I think it would be better to mark all files except the entry file as external, since bundling them together skews code coverage...

And, do you want me to update the Readme?

ambar commented:
@cpiber Thanks for the PR! You're right, maybe it's simpler to exclude entry file.

ambar commented on test/__snapshots__/transform.spec.js.snap:
```diff
@@ -29,3 +29,12 @@ Object.defineProperty(exports, 'URL', {
+exports[`process should transform relative imports 1`] = `
+"'use strict';
+
+const foo = 42;
```
Bundled code will break the code coverage, imported files should remain external.

cpiber replied:
Yes, hence my question ;)

It does have the advantage of speed however, so maybe include it as an optional feature?

cpiber replied:
Also you transpile the config file, there it definitely makes sense to include all imports

cpiber replied:
I implemented now that by default, no imports are resolved, except for the config, where relative imports are resolved. It is controlled by an option.

ambar commented on transform.js:
```diff
@@ -27,7 +27,7 @@ const external = () => {
     resolveId(id) {
       // filter relative or absolute imported modules
-      if (builtins.includes(id) || !id.startsWith('.') || !isAbsolute(id)) {
+      if (builtins.includes(id) || (!id.startsWith('.') && !isAbsolute(id))) {
```
Relative file should be external, and it's also used to suppress the console warning:

```
➜  rollup-jest git:(pr/13) ✗ yarn jest --no-cache
yarn run v1.22.10
'..' is imported by test/transform.spec.js, but could not be resolved – treating it as an external dependency
 PASS  test/main.spec.js
 PASS  test/transform.spec.js
```

cpiber replied:
Your code here suggested that you wanted to exclude relative files. In reality, we don't want to import anything, so we should just return false here. Then we could combine that with the first plugin.

cpiber commented:
Also might want to look at https://github.com/facebook/jest/tree/master/packages/jest-transform. ~~There's not much documentation but it should help with setup, and it provides a way of asynchronous compilation~~ (Never mind, seems it's only declarations and helpers for babel-jest, not really useful... I added the async options, but they're not used in current versions of jest afaik). Though that'd mean setting a minimum version of jest (which you should do anyway, as a peer dependency)...

~~In any case, a source map and cache key would also be nice, jest supports them~~: https://jestjs.io/docs/code-transformation (added that to PR)

cpiber committed:
Sourcemap support

cpiber committed:
Cache keys, async for future

cpiber committed:
Enable sourcemaps, dont resolve imports by default

cpiber committed:
Allow for only relative resolve and use in config

cpiber committed:
Readme

cpiber committed:
Fix cache keys called with wrong options object

cpiber committed:
Include full config in cache key

This should avoid the same file being used e.g. with and without instrumentation

ambar commented:
@cpiber Released as v1.1.2, thanks for the update.

cpiber commented:
Thanks to you for the package :grin:
